### PR TITLE
Use Metadata to provide improved JSON Schemas in OpenAPI

### PR DESCRIPTION
**Pull request by bauer-andreas**
# Use Metadata to provide improved JSON Schemas in OpenAPI

## :recycle: Current situation & Problem

Currently the JSONSchema for Request, Response and Parameter descriptions are solely sourced from static type information. The user has no possibility to adjust or manually modify the resulting scheme.

## :bulb: Proposed solution
This PR introduces a generalised `JSONSchemeModification` system which can be used, with an according ContextKey, to provide modifications to a schema. The JSONSchemeModification system is target internal.

Means of modifying the schema are provided by according Content Metadata Definitions. By building upon this generalised system, it is easy to add further Metadata Definitions in the future.

## :gear: Release Notes 

This PR adds the following new Metadata to the `ApodiniOpenAPI` target:

### Handler Metadata
* `Summary(_:)`

### Component Metadata
* `ParameterDescription(for:_:)`
* `Tags(_:)` (this complements the already existent `.tags` modifier)

### Content Metadata
Each of those Metadata can be declared on the Metadata declaration blocks of `Content` types.
Typically, each of those Metadata provide a initilizer to apply to the `Content` type itself or to apply to one of its properties.
As parameter name can't be derived from Swift KeyPaths, you always need to supply both: a KeyPath to the property and the type name in string representation.

* **Applying for all Schemas**:
  * `Description(_:)` or `Description(of:_:propertyName:)`
  * `MarkDeprecated(_:)` or `MarkDeprecated(property:_:propertyName:)`
  * `Example(_:)` or `Example(for:_:propertyName:)`
* **Applying to Numeric and Integer Schemas**:
  * `Maximum(of:is:exclusive:propertyName:)`
  * `Minimum(of:is:exclusive:propertyName:)`
  * `MultipleOf(of:is:propertyName:)`
* **Applying to String Schemas**:
  * `MaxLength(of:is:propertyName:)`
  * `MinLength(of:is:propertyName:)`
  * `Pattern(of:is:propertyName:)`

### Related PRs
- #232 

### Testing
Extensive testing was added.

### Reviewer Nudging
Core of the system is the `JSONSchemeModification`. This was needed to modify `JSONScheme.XXXContext` values (which are not built to be mutable). All the Metadata then only builds around this modification system.

### CODE REVIEW

**Comment by bauer-andreas** [comment]
ping to the reviewers 🥲

**Comment by philippzagar** [comment]
So sorry @Supereg, I completely missed this PR. 
At the moment I'm a little bit swamped with writing my thesis. Maybe @hendesi can do the review? :)

**Comment by hendesi** [comment]
Oh sh*t! Sorry completely forgot about this!
I will give this a look today 👍 Sorry for the delay

File: Sources/ApodiniOpenAPI/Extensions/AnyCodable+Complex.swift [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        }
+
+        let json: Any
+        do {
+            json = try JSONSerialization.jsonObject(with: encoded, options: [])
+        } catch {
+            fatalError("Failed to serialize json encoded string: \(String(data: encoded, encoding: .utf8) ?? "<ERR>") [\(element)]: \(error)")
+        }
+
+        #if DEBUG
```
Comment (hendesi) [inline Sources/ApodiniOpenAPI/Extensions/AnyCodable+Complex.swift]: What's reasoning behind this being debug only?

**Review by hendesi — APPROVED** [review]
To my very limited understanding of the metadata internals, this looks good to me. 👍 
I just had one small comment/question, but that should be fairly simple, so I'd already approve it.
Sorry for blocking this so long

File: Sources/ApodiniOpenAPI/Extensions/AnyCodable+Complex.swift (hunk 1 above)
Comment (bauer-andreas) [reply Sources/ApodiniOpenAPI/Extensions/AnyCodable+Complex.swift]: Oh. That's a forgotten artifact from development actually. Good you pointed that out.

**Pull request merged**